ext/test/zpages/tracez_processor_test.cc

```diff
@@ -0,0 +1,629 @@
+  EXPECT_TRUE(ContainsNames(span_names, completed, 1, 3, true)); // s2 s1
+  EXPECT_EQ(running.size(), 2);
+  EXPECT_EQ(completed.size(), 2);
+
+  // End remaining Spans
+  span_vars[0]->End();
+  span_vars[4]->End();
+  UpdateSpans(processor, completed, running, true);
+
+  EXPECT_TRUE(ContainsNames(span_names, completed, 0, 1)); // s0
+  EXPECT_TRUE(ContainsNames(span_names, completed, 4)); // s
+  EXPECT_EQ(running.size(), 0);
+  EXPECT_EQ(completed.size(), 2);
+
+  UpdateSpans(processor, completed, running, true);
+
+  EXPECT_EQ(running.size(), 0);
+  EXPECT_EQ(completed.size(), 0);
+}
+
+
+/*
+ * Test if multiple spans move from running to completed at expected times,
+ * running/completed spans are split. Ensure correct behavior even when
+ * multiple spans are discarded, even when span starting and ending is
+ * non-sequential. Only new completed spans are stored.
+*/
+TEST(TracezSpanProcessor, MultipleSpansOuterSplitNewOnly) {
+  std::shared_ptr<TracezSpanProcessor> processor(new TracezSpanProcessor());
+  auto tracer = std::shared_ptr<opentelemetry::trace::Tracer>(new Tracer(processor));
+  auto spans = processor->GetSpanSnapshot();
+  auto running = spans.running;
+  auto completed = std::move(spans.completed);
+
+  std::vector<std::string> span_names = {"s0", "s2", "s1", "s1", "s"};
+
+  std::vector<opentelemetry::nostd::unique_ptr<opentelemetry::trace::Span>> span_vars;
+  for (const auto &name : span_names) span_vars.push_back(tracer->StartSpan(name));
+
+  // End last span
+  span_vars[4]->End();
+  UpdateSpans(processor, completed, running, true);
+
+  EXPECT_TRUE(ContainsNames(span_names, running, 0, 4, true)); // s0 s2 s1 s1
+  EXPECT_TRUE(ContainsNames(span_names, completed, 4, 5, true)); // s
+  EXPECT_EQ(running.size(), 4);
+  EXPECT_EQ(completed.size(), 1);
+
+  // End first span
+  span_vars[0]->End();
+  UpdateSpans(processor, completed, running, true);
+
+  EXPECT_TRUE(ContainsNames(span_names, running, 1, 4, true)); // s2 s1 s1
+  EXPECT_TRUE(ContainsNames(span_names, completed, 0, 1, true)); // s0
+  EXPECT_EQ(running.size(), 3);
+  EXPECT_EQ(completed.size(), 1);
+
+  // End remaining middle pans
+  for (int i = 1; i < 4; i++) span_vars[i]->End();
+  UpdateSpans(processor, completed, running, true);
+
+  EXPECT_TRUE(ContainsNames(span_names, completed, 1, 4, true)); // s2 s1 s1
+  EXPECT_EQ(running.size(), 0);
+  EXPECT_EQ(completed.size(), 3);
+
+  UpdateSpans(processor, completed, running, true);
+
+  EXPECT_EQ(running.size(), 0);
+  EXPECT_EQ(completed.size(), 0);
+}
+
+
+/*
+ * Test if thread safety is compromised and errors are thrown when many spans
+ * start at the same time. Test is not failed if no errors occur.
+ */
+TEST(TracezSpanProcessor, RunningThreadSafety) {
+  std::shared_ptr<TracezSpanProcessor> processor(new TracezSpanProcessor());
+  auto tracer = std::shared_ptr<opentelemetry::trace::Tracer>(new Tracer(processor));
+  std::vector<opentelemetry::nostd::unique_ptr<opentelemetry::trace::Span>> spans;
```

Comment:
You also have to lock around concurrent writes/reads to this vector (also in all other tests).

As far as I can see you application code is thread safe, but the test code is not.

Reply:
Gotcha, thanks for the catch! I changed the functions to use separate vectors, so there's no unsafe writes/reads and added locking if that works with you